data/Builder.go

```diff
@@ -0,0 +1,110 @@
+package data
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+//Datum represent one data point
+type Datum map[string]interface{}
+
+//Dataset represents an array of data points
+type Dataset []Datum
+
+//GetSelector returns the selector fields for a given platform datatype, or nil if there is no match
+func GetSelector(data interface{}) interface{} {
+
+	switch data.(type) {
+	case *Basal:
+		return data.(*Basal).Selector()
+	case *DeviceEvent:
+		return data.(*DeviceEvent).Selector()
+	default:
+		return nil
+	}
+}
+
+//Builder interface that the TypeBuilder implements
+type Builder interface {
+	BuildFromRaw(raw []byte) (interface{}, *Error)
+	BuildFromData(data map[string]interface{}) (interface{}, *Error)
+	BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet)
+}
+
+//TypeBuilder that is used to build data types that the platform understands
+type TypeBuilder struct {
+	inject map[string]interface{}
+}
+
+//NewTypeBuilder returns an instance of TypeBuilder
+func NewTypeBuilder(inject map[string]interface{}) Builder {
+	return &TypeBuilder{
+		inject: inject,
+	}
+}
+
+//BuildFromDataSet will build the matching type(s) from the given Dataset
+func (typeBuilder *TypeBuilder) BuildFromDataSet(dataSet Dataset) ([]interface{}, *ErrorSet) {
+
+	var set []interface{}
+	var buildError *ErrorSet
+
+	for i := range dataSet {
+		item, err := typeBuilder.BuildFromData(dataSet[i])
+		if err != nil && !err.IsEmpty() {
+			if buildError == nil {
+				buildError = NewErrorSet()
+			}
+			buildError.AppendError(err)
+			continue
+		}
+		set = append(set, item)
```

Comment:
If we have an error, then an item will not be placed into the final array and the incoming array and outgoing array will not be coordinated.  What if we actually just stored the error back into the array after building.  So, the outgoing array would be an array of both built data as well as errors?